lib/mandrill.rb

```diff
@@ -29,13 +29,19 @@ def initialize(apikey=nil, debug=false)
             @apikey = apikey
         end
 
+        MAX_RETRY = 3
+
         def call(url, params={})
-            params[:key] = @apikey
-            params = JSON.generate(params)
-            r = @session.post(:path => "#{@path}#{url}.json", :headers => {'Content-Type' => 'application/json'}, :body => params)
-            
-            cast_error(r.body) if r.status != 200
-            return JSON.parse(r.body)
+          retry_count ||= 0
+          params[:key] = @apikey
+          params = JSON.generate(params)
+          r = @session.post(:path => "#{@path}#{url}.json", :headers => {'Content-Type' => 'application/json'}, :body => params)
+          JSON.parse(r.body)
```

Comment:
i think we'll want to keep the `cast_error(r.body) if r.status != 200` here (above `JSON.parse(r.body)`) so that we can `rescue` from it

Reply:
i thought we enter the rescue block from the post request (line 38), and that `cast_error` should only be called if we've hit max retry attempts and `r.status != 200`

so it'd look like this:

```
        def call(url, params={})
          retry_count ||= 0
          params[:key] = @apikey
          params = JSON.generate(params)
          r = @session.post(:path => "#{@path}#{url}.json", :headers => {'Content-Type' => 'application/json'}, :body => params)
          JSON.parse(r.body)
        rescue
          cast_error(r.body) if retry_count >= MAX_RETRY
          sleep 2**retry_count
          retry_count += 1
          retry
        end

Reply:
the original error that was raised (we see it output in slack #app-messages because of the sidekiq error_handlers config in sidekiq.rb):

> We received an unexpected error: <html><body><h1>504 Gateway Time-out</h1>
The server didn't respond in time.
</body></html>

looks a lot like it came from `raise Error, "We received an unexpected error: #{body}"` in `cast_error`. i don't think the `post` would commonlky raise errors, instead it's the `parse` calls

Reply:
ahhhh i see. thanks for explaining.